### PR TITLE
Foreach adapt

### PR DESCRIPTION
**Pull request by gonidelis**
Adapt for_each in order to support for the begin and end iterators to be of different type

### CODE REVIEW

File: libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp [hunk 1]
```diff
@@ -26,6 +26,7 @@
 #include <hpx/execution/algorithms/detail/is_negative.hpp>
 #include <hpx/executors/execution_policy.hpp>
 #include <hpx/parallel/algorithms/detail/dispatch.hpp>
```
Comment (gonidelis) [inline libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp:28]: Is this needed?

Comment (hkaiser) [reply libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp:28]: I think it's needed. This is the base class of the algorithm implementation (see `detail::for_each` below)

File: libs/algorithms/tests/unit/container_algorithms/foreach_adapt.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#include <hpx/hpx_main.hpp>
```
Comment (hkaiser) [inline libs/algorithms/tests/unit/container_algorithms/foreach_adapt.cpp:1]: Please add copyright, license, and SPDX tag.

Comment (gonidelis) [reply libs/algorithms/tests/unit/container_algorithms/foreach_adapt.cpp:7]: Sorry, missed it. Just fixed it and clang-formatted the code.

**Comment by gonidelis** [comment]
Should I use the `is_sentinel_for` trait too?

File: libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp [hunk 3]
```diff
@@ -224,10 +224,10 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
         }
         // forward declare the segmented version of for_each_ algorithm
-        template <typename ExPolicy, typename SegIter, typename F,
-            typename Proj>
-        inline typename util::detail::algorithm_result<ExPolicy, SegIter>::type
-        for_each_(ExPolicy&& policy, SegIter first, SegIter last, F&& f,
+        template <typename ExPolicy, typename SegIterB, typename SegIterE,
+            typename F, typename Proj>
+        inline typename util::detail::algorithm_result<ExPolicy, SegIterB>::type
+        for_each_(ExPolicy&& policy, SegIterB first, SegIterE last, F&& f,
```
Comment (hkaiser) [inline libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp:230]: You might want to change the implementation of this function as well (just the prototype) ...

Comment (gonidelis) [reply libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp:230]: @hkaiser Isn't that the prototype? What other part should I change?

Comment (hkaiser) [reply libs/algorithms/include/hpx/parallel/algorithms/for_each.hpp:230]: See here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/segmented_algorithms/include/hpx/parallel/segmented_algorithms/for_each.hpp#L183-L187

**Comment by msimberg** [comment]
add to whitelist

**Comment by gonidelis** [comment]
**Important Note**: Please don't merge yet. Need to add a couple of doxygen comments.

File: libs/segmented_algorithms/include/hpx/parallel/segmented_algorithms/for_each.hpp [hunk 4]
```diff
@@ -180,23 +180,23 @@ namespace hpx { namespace parallel { inline namespace v1 {
...
         {
             typedef parallel::execution::is_sequenced_execution_policy<ExPolicy>
                 is_seq;
 
             if (first == last)
             {
-                typedef util::detail::algorithm_result<ExPolicy, SegIter>
+                typedef util::detail::algorithm_result<ExPolicy, SegIterB>
                     result;
                 return result::get(std::move(last));
```
Comment (hkaiser) [inline libs/segmented_algorithms/include/hpx/parallel/segmented_algorithms/for_each.hpp:196]: You might want to return `first` here.

**Pull request merged**